lib/Doctrine/DBAL/Schema/Column.php

```diff
@@ -118,9 +119,17 @@ public function setOptions(array $options)
     {
         foreach ($options as $name => $value) {
             $method = "set".$name;
-            if (method_exists($this, $method)) {
-                $this->$method($value);
+            if ( ! method_exists($this, $method)) {
+                // next major: throw an exception
+                @trigger_error(sprintf(
```

Comment:
The thread started in https://github.com/doctrine/dbal/pull/2846#discussion_r155290648.

> most implementations ignore silenced errors, this one does not

That's my point. Why should this error be ignored by most implementations?

> If we don't users, will get a deprecation *and won't be able to do anything about it*.

If we trigger suppressed errors, the users will not see any errors and won't be able to react unless their code is covered by tests which uses the same listener.

Reply:
> If we trigger suppressed errors, the users will not see any errors and won't be able to react unless their code is covered by tests which uses the same listener.

Yes, and that's anyone that uses the phpunit-bridge component (so not just sf users, but also any library that relies on it for testing), but anyone who uses that particular component. That's not the only advantage, see https://github.com/doctrine/dbal/pull/2846#discussion_r155290648

Here is the list of the > 1200 libs that use it: https://packagist.org/packages/symfony/phpunit-bridge/dependents

If you are still not convinced, then maybe this should wait for a PSR an even larger portion of the population benefits from that.

Reply:
> Yes, and that's anyone that uses the phpunit-bridge component (so not just sf users, but also any library that relies on it for testing)

This is fine. But there are also libraries which don't use the bridge component. Why should we suppress this error for them?

> We want it here not only for Symfony users, but for us: we want to know if we use a deprecated path somewhere else in the lib

I totally get it. I'm asking why we need to suppress the error, not why we need to trigger it.

> Here is a list of the > 1200 libs that use it: https://packagist.org/packages/symfony/phpunit-bridge/dependents

It doesn't explain why the error needs to be suppressed.

> If you are still not convinced, then maybe this should wait for a PSR an even larger portion of the population benefits from that.

You don't have to convince me. I'm just keeping asking the same question which I'd like to get an answer for.

Reply:
> This is fine. But there are also libraries which don't use the bridge component. Why should we suppress this error for them?

Because otherwise I'm afraid their error handler will turn it into an exception.

Reply:
So not silencing would not be BC at all :P

Reply:
> Because otherwise I'm afraid their error handler will turn it into an exception.

As far as I can tell, most error handlers override the way how the errors are _logged_ (e.g. instead of displaying them, log them to a logging service). If the handler converts errors to exceptions, it should understand that a deprecated error is not an exceptional situation. Logically, its severity is even lower than a notice because the code keeps working as designed.

I see what you're saying but if we want to inform users about them using the deprecated APIs, this message should be available for them without additional means (the bridge). Otherwise, we will not inform the users who don't use the bridge (which is fine) in favor of those who have badly designed error-to-exception converters (which is not).

Reply:
> So not silencing would not be BC at all :P

It depends on how BC is defined. If "not being informed about using deprecated APIs" is part of the previous API, then yes. Was the previous API defined like this?

Reply:
> Was the previous API defined like this?

AFAIK it wasn't defined at all, which should probably mean we can't suddenly decide people should be informed through an error, moreover it it might be turned into an exception. 

> I see what you're saying but if we want to inform users about them using the deprecated APIs, this message should be available for them without additional means (the bridge). Otherwise, we will not inform the users who don't use the bridge (which is fine) in favor of those who have badly designed error-to-exception converters (which is not).

I don't know any other dynamic way. There are static ways like the `@deprecated` tag, but that's it.

> in favor of those who have badly designed error-to-exception converters (which is not).

Sorry, I'm a bit slow, I didn't quite understand that last bit. Do you mean error to exception converters who would also ignore the silence operator?

Reply:
> [...] we can't suddenly decide people should be informed through an error.

Why not? It's a non-fatal error. If triggering an error is a bad idea why do we want to use `trigger_error()`?

> [...] moreover it it might be turned into an exception

How is it expected to happen?

> Do you mean error to exception converters who would also ignore the silence operator?

No, I mean the converters who convert deprecated errors to exceptions. There's nothing exceptional if deprecated code is used in production.

Reply:
> Why not? It's a non-fatal error. If triggering an error is a bad idea why do we want to use trigger_error()?

I think that would be because how php behaves if you have no error handler at all:

https://3v4l.org/7WsRc

IMO this should probably something in the error_log, but not display anything.

> How is it expected to happen?

I think having both errors and exceptions in the php language is not great, and that flaw started being addressed with php 5.1 with this [ErrorException](http://php.net/manual/en/class.errorexception.php), and since then it's pretty common practice for error handlers to do that, here is a random, non-symfony example to back that up: https://github.com/filp/whoops/blob/2abce9d956589122c6443d6265f01cf7e9388e3c/src/Whoops/Run.php#L349 . So I really think it's a sensible assumptionto make.

> No, I mean the converters who convert deprecated errors to exceptions. There's nothing exceptional if deprecated code is used in production.

I would call that "in favor of", I would call that a big BC-break. So maybe it should first be introduced with the silence operator, and then the silence operator should be removed on next major? But that wouldn't make much sense since the deprecated paths would be removed as well.


Here is my source for saying this could become a PSR: https://medium.com/php-fig/a-month-of-php-fig-1-october-2017-af253682ef5b

Reply:
> I think that would be because how php behaves if you have no error handler at all:
> 
> https://3v4l.org/7WsRc

This is exactly how it should work. It will let developers know they are using deprecated APIs without having to use any additional libraries.

> IMO this should probably produce something in the error_log, but not display anything.

It should be controlled by `display_errors` and `error_log` INI settings, not by the DBAL code.

> here is a random, non-symfony example to back that up

Why is a deprecated error an exception? At runtime, there's nothing exceptional about using deprecated APIs.

> I would call that "in favor of", I would call that a big BC-break.

What exactly is being broken?

> Here is my source for saying this could become a PSR

Could you pls quote what could become a PSR and point to a document describing the proposal?

Reply:
> This is exactly how it should work. It will let developers know they are using deprecated APIs without having to use any additional libraries.

What about test suite of other projects that don't use any error handler? Is that the case for `doctrine/orm` for instance?


> What exactly is being broken?

upgrading the lib produces something on the output. Let me ask the sf community about this, maybe I didn't grasp it entirely.



> Could you pls quote what could become a PSR and point to a document describing the proposal?

That's all the details I have right now: 

> For handling deprecations of classes and methods to trigger errors/notifications

Reply:
@morozov I think I found the real reason:

> to guarantee opt-in of deprecation. Otherwise, it's only opt-in with a "reasonable error handler", which isn't that opt in. Besides, the default php.ini production value for `error_reporting` only hides PHP deprecations, not user deprecations

— @wouterjnl

So I guess silencing is up to you, and I will remove it if you want me to.

Reply:
@morozov as mentioned here and elsewhere, the appropriate way to trigger deprecation notices is `@trigger_error(..., E_USER_DEPRECATED);`

As explained by @nicolas-grekas, without the error suppression operator the code would start spitting out deprecation warnings to STDERR which we don't want: deprecation notices are an opt-in feature: users should not be bothered by it unless they specifically decide to care. Thus, the only way to solve this is to suppress the `trigger_error` call and let a dedicated deprecation handler (such as the listener from PhpunitBridge) log the deprecations and display them accordingly.

Reply:
Thanks for providing a clear explanation, I had a hard time doing it!

Reply:
@greg0ire sorry for giving you a hard time :smile:

Reply:
No problem! It was worth it. Very reassuring to see people are doing due diligence on that project!